### ISSUE TITLE
loss computation with mask has a bug

### ISSUE THREAD

**Issue by udibr** [fchollet/keras#2546]
https://github.com/fchollet/keras/pull/2541

If you have for example time series output with a mask then current code multiply the loss by
1/mean(mask)
https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L319
but this does not make much sense because different batches will have different scaling
and as a result a higher weight in training will be given to batches that happens by accident to have many short samples (i.e. a lot of zeros in their mask)
this also scales the reported loss by the same factor while you would expect the loss to be just the sum over all steps.

instead what you should do is first find out if any of the samples in the batch was unmasked.
In the case of time series this will happen in the rare case in which all the steps of a single sample had all mask elements set to 0.

you should then compute the mean of the entire batch but only on the samples which were masked.
For example by factoring by `1/mean(per_sample_mask)`
In most cases this will always be all samples so this bug had very little effect with this respect.

However the usage of mean in the past over `mask` entered a single batch scaling factor which is usually was > 1 so after the PR your loss values should drop by a big factor

**Comment by braingineer** [comment]
I thought this once too. It turns out, no.  I had to do the math to convince myself, because if you test it empirically, it turns out to be the exact same (try it yourself). 

I'm not sure about the weighted batch because I haven't done the math for that one, but, the math for the non weighted:

```python
#score_array = [ [ x_ij , ... ] , ...]
# zero out the masked things
score_array *= mask
M = K.mean(mask)
score_array /= M
# now, score array is [ [x_ij / M , .. ] ...]
#then when you take the mean of score array, 
K.mean(score_array)
# what you are doing is 
# sum([ [ x_ij / M , ... ] .. ]) /  N 
# but you want 
# sum([[x_ij, ...]]) / sum(mask)  (sum of mask is the number of nonzero things 
# so, pull out the M
# 1/M * sum(X) / N 
# sum(X) / (M * N)
# sum(X) / (sum(mask) / N * N)
# sum(X) / sum(mask)
```

Now, if you're using the accuracy metric, you do have an issue.  It should be the following, but with the appropriate tensor flow operation. 

```python
def categorical_accuracy(y_true, y_pred, mask=None):
    if mask is not None:
        if K.ndim(y_pred) == K.ndim(mask):
            mask = K.max(mask, axis=-1)
        eval_shape = (reduce(mul, K.shape(y_true)[:-1]), K.shape(y_true)[-1])
        y_true_flat = K.reshape(y_true, eval_shape)
        y_pred_flat = K.reshape(y_pred, eval_shape)
        mask_flat = K.flatten(mask).nonzero()[0]  ### how do you do this on tensorflow?
        evalled = K.gather(K.equal(K.argmax(y_true_flat, axis=-1),
                                   K.argmax(y_pred_flat, axis=-1)), 
                           mask_flat)
        return K.mean(evalled)
```

**Comment by udibr** [comment]
Thanks but I'm talking about a situation in which each sample has a different mask.
I think (and of course I may be wrong) that the loss in this case should be to average over samples (if they are not unmasked)
and that the loss for each sample is to sum the losses over the masked time steps and not to take the mean over the masked time step.

the loss of each sample should be the log-likelihood of the sample and the LK of each sample is the sum of the LK of each step within the sample (which is masked)
as you can see you don't need to divide by the number of steps.

if you do divide by number of steps (sum of the mask bits) then what happens is that you penalize the LK of longer samples and as a result your model will prefer to generate shorter outputs.

**Comment by braingineer** [comment]
> each sample has a different mask
I don't understand.  The math I just showed was agnostic to that situation. 

The mask could be
```
[ 1 1 0 0
  1 0 0 0
  1 1 1 0 ]
```

Code:
```python
A = np.arange(1,31).reshape(6,5).astype(np.float32)
mask = np.zeros((6,5))
for i in range(6):
    ind = np.random.randint(1, 4)
    mask[i, :ind] = 1
A *= mask
A = np.exp(A) / np.sum(np.exp(A), axis=-1, keepdims=True) 
print(np.mean(A) / np.mean(mask))
print(np.mean(A/np.mean(mask)))
```

**Comment by udibr** [comment]
I closed this issue/PR because I realized that the industry standard is to do what the current code does.
the standard is to minimize the [perplexity](https://www.tensorflow.org/versions/r0.8/tutorials/recurrent/index.html#loss-function)
which is the per-word loss.

Using the correct loss function for your specific needs is a big issue in ML which is usually overlooked in favour of using a standard loss.

I think that if you are interested in building RNN for language then you can ask what is the loss of an entire sentence not just the loss of an average word in the sentence. of course longer sentences will have more words in them as a result will tend to have higher loss

**Comment by braingineer** [comment]
It makes no difference udibr.   You will find greater numerical stability by using the mean. Additionally, the mean of the log probability is the empirical cross entropy, which is heavily rooted in information theory.  There is a great amount of background literature on this topic.  

Though, It's actually kind of funny---they define perplexity differently on that page that the people who have been using it for decades.  Perplexity has historically be 2 ** H(x)  with the cross entropy using the base 2 log.

**Comment by braingineer** [comment]
Here is a couple of the early papers for you:

http://scitation.aip.org/content/asa/journal/jasa/62/S1/10.1121/1.2016299

http://citeseer.ist.psu.edu/viewdoc/summary?doi=10.1.1.13.9919

**Comment by udibr** [comment]
thanks
